api/explorer/api/queries/packages.py

```diff
@@ -11,20 +11,112 @@
 
 logger = logging.getLogger(__name__)
 
+# Since the provided search predicate filters by name, we need to know the name
+# of the relevant vertex property. There might be smarter ways to extract this
+# rather than hard-coding it here.
+NAME_PROPERTY = "pname"
+
 
-# @dataclass
 class ListPackagesRequest(BaseModel):
     cursor: Optional[Cursor] = None
     search_predicate: Optional[str] = None
     limit: int = 10
 
 
-# @dataclass
 class ListPackagesResponse(BaseModel):
     new_cursor: Optional[Cursor]
     packages: list[Package]
 
 
+class ListUniquePackageNamesRequest(BaseModel):
+    cursor: Optional[Cursor] = None
+    search_predicate: Optional[str] = None
+    limit: int = 10
+
+
+class ListUniquePackageNamesResponse(BaseModel):
+    new_cursor: Optional[Cursor]
+    package_names: list[str]
+
+
+def list_unique_package_names(
+    request: ListUniquePackageNamesRequest, conn: DriverRemoteConnection
+) -> ListUniquePackageNamesResponse:
+    """Lists the unique names of packages available in the specified Gremlin Server
+
+    Args:
+        request (ListUniquePackageNamesRequest): The request parameters
+        conn (DriverRemoteConnection): A Gremlin Driver remote connection
+
+    Returns:
+        ListUniquePackageNamesResponse: A new cursor (if available) and the
+        list of package names matching the provided request.
+    """
+
+    # Connect to Gremlin Server and start a new traversal
+    g = traversal().withRemote(conn)
+
+    # First filter traversal based on predicate if one exists
+    filtered_traversal = g.V().values(NAME_PROPERTY)
+    if request.search_predicate:
+        logger.debug("Filtering traversal using provided predicate.")
+        filtered_traversal = filtered_traversal.is_(
+            TextP.startingWith(request.search_predicate)
+        )
+
+    # Apply ordering to the traversal
+    if not request.cursor or request.cursor.direction is CursorDirection.NEXT:
+        ordering = Order.asc  # type: ignore
+    else:
+        ordering = Order.desc  # type: ignore
+    ordered_filtered_traversal = filtered_traversal.order().by(ordering)
+
+    # Then we can filter based on the cursor, if one exists.
+    if request.cursor:
+        if request.cursor.direction == CursorDirection.NEXT:
+            logger.debug("cursor direction=next")
+            cursor_predicate = TextP.gte(request.cursor.row_id)
+        else:
+            logger.debug("cursor direction=previous")
+            cursor_predicate = TextP.lte(request.cursor.row_id)
+        page_traversal = (
+            ordered_filtered_traversal.is_(cursor_predicate)
+            .dedup()
```

Comment:
I guess this `dedup` is the core of the change here?

Reply:
It's one of the main changes. The other being that the output of the traversal is a `values` step rather than an `elementMap`. This also requires a bit of a change to the filtering logic as well since you need to use an `is` step since the traversal type is `string` rather than `vertex`.